mchf-eclipse/drivers/audio/psk.c

```diff
@@ -9,10 +9,63 @@
 
 // Common
 #include "psk.h"
-#include "ui_driver.h"
-#include "radio_management.h"
+#include <math.h>
 #include <stdlib.h>
+#include "softdds.h"
 #include "uhsdr_digi_buffer.h"
+#include "ui_driver.h" // only necessary because of UiDriver_TextMsgPutChar
+#include "radio_management.h" // only necessary because of RadioManagement_Request_TxOff
```

Comment:
I would like to handle it by adding callbacks for all digimodes and add an interface to register these Callbacks:  TxRequest and RxRequest....  What do you think?

Reply:
I don't understand what is your idea. Who is registering and who is calling the call backs?

A few comments:
The Tx Off/On Request functions just sets a flag, which tells the control logic to switch back to receive (or  transmit) at a convenient time. You can call this function anytime anywhere. 
I think, it is safe to say, we cannot decouple things much more than that.

If you mean to identify a common interface which all digi modes are supposed to implement, defining a structure of function pointers serving as interface to the modes (similar to what we do with the oscillators), well, that is possible and might be useful. 

I am fully into trying to identify common patterns, aligning signatures of functions serving a similar purpose, extracting shared code into separate functions in order to reduce code duplication and lower the maintenance effort where it makes sense, etc. All of this is a prerequisite for doing this kind of generalization. This was part of my motivation when I extracted the tx_processor code and refactored the code. However, looking at the code as it is, I see code duplication happening if we decouple things "too much" this way. 

I did the analysis which led to the two comments above mostly to understand why and how ui_driver is used here. Which should not be necessary. However, factoring out the buffer out of ui_driver is sufficient to solve the problem. A buffer like this should be exactly that, a buffer between different parts of the system which exchange data through the buffer asynchronously.

Since I don't have an answer for my initial question, I have not idea if your callbacks are something I am interested in or not. Generally I am not a too big of a fan for dynamically coupling everything since it makes navigating in the source code very difficult since you have to "execute" the dynamic registering in your head if you want to follow the path of execution just by looking at the source. 

That human execution is error-prone and takes more time to grasp than a direct call.  So at component boundaries it may make sense.